npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -484,6 +440,27 @@ func (iMgr *IPSetManager) exists(name string) bool {
 	return ok
 }
 
+// the metric for number of ipsets in the kernel will be lower than in reality until the next applyIPSet call
+func (iMgr *IPSetManager) modifyCacheForCacheDeletion(set *IPSet, deleteOption util.DeleteOption) {
+	if deleteOption == util.ForceDelete {
+		// If force delete, then check if Set is used by other set or network policy
+		// else delete the set even if it has members
+		if !set.canBeForceDeleted() {
+			return
+		}
+	} else if !set.canBeDeleted() {
+		return
+	}
+
+	delete(iMgr.setMap, set.Name)
+	metrics.DecNumIPSets()
+	if iMgr.iMgrCfg.IPSetMode == ApplyAllIPSets {
+		// NOTE: in ApplyAllIPSets mode, if this ipset has never been created in the kernel, it would be added to the deleteCache, and then the OS would fail to delete it
+		iMgr.modifyCacheForKernelRemoval(set.Name)
+	}
+	// if mode is ApplyOnNeed, the set will not be in the kernel (or will be in the delete cache already) since there are no references
```

Comment:
Even in applyONNeed we might end in situation where a set is needed with members and then we delete the referenced netpol/list. We can end up with empty sets in kernel right ?

Reply:
anytime the set is modified, we see if it should be in the kernel now, and it to the deleteCache if needed